specs/merge/beacon-chain.md

```diff
@@ -0,0 +1,257 @@
+```
+
+#### `get_evm_beacon_block_roots`
+
+```python
+def get_evm_beacon_block_roots(state: BeaconState) -> Sequence[Bytes32]:
+    num_block_roots = min(EVM_BLOCK_ROOTS_SIZE, SLOTS_PER_HISTORICAL_ROOT)
+    return get_recent_beacon_block_roots(state, num_block_roots)
+```
+
+### Block processing
+
+```python
+def process_block(state: BeaconState, block: BeaconBlock) -> None:
+    process_block_header(state, block)
+    process_randao(state, block.body)
+    process_eth1_data(state, block.body)  # [Modified in The Merge]
+    process_operations(state, block.body)
+    process_application_payload(state, block.body)  # [New in The Merge]
+```
+
+#### Modified `process_eth1_data`
+
+*Note*: The function `process_eth1_data` is modified to update `state.eth1_data` with `eth1_data` of each block.
+
+```python
+def process_eth1_data(state: BeaconState, body: BeaconBlockBody) -> None:
+    state.eth1_data = body.eth1_data
+```
+
+#### Application payload processing
+
+##### `BeaconChainData`
+
+*Note*: `BeaconChainData` contains beacon state data that is used by the application state transition function.
+
+```python
+class BeaconChainData(Container):
+    slot: Slot
+    randao_mix: Bytes32
+    timestamp: uint64
+    recent_block_roots: Sequence[Bytes32]
+```
+
+##### `get_application_state`
+
+*Note*: `ApplicationState` class is an abstract class representing ethereum application state.
+
+Let `get_application_state(application_state_root: Bytes32) -> ApplicationState`  be the function that given the root hash returns a copy of ethereum application state. 
+The body of the function is implementation dependant.
+
+##### `application_state_transition`
+
+Let `application_state_transition(application_state: ApplicationState, beacon_chain_data: BeaconChainData, application_payload: ApplicationPayload) -> None` be the transition function of ethereum application state. 
+The body of the function is implementation dependant.
+
+*Note*: `application_state_transition` must throw `AssertionError` if either the transition itself or one of the post-transition verifications has failed.
+
+##### `process_application_payload`
+
+```python
+def process_application_payload(state: BeaconState, body: BeaconBlockBody) -> None:
+    """
+    Note: This function is designed to be able to be run in parallel with 
+    the other `process_block` sub-functions
+    """
+    
+    # Utilizes `compute_randao_mix` to avoid any assumptions about
+    # the processing of other `process_block` sub-functions
+    beacon_chain_data = BeaconChainData(
+        slot=state.slot,
+        randao_mix=compute_randao_mix(state, body.randao_reveal),
+        timestamp=compute_time_at_slot(state.genesis_time, state.slot),
+        recent_block_roots=get_evm_beacon_block_roots(state) 
+    )
+    
+    application_state = get_application_state(state.application_state_root)
+    application_state_transition(application_state, beacon_chain_data, body.application_payload)
```

Comment:
Is the beacon node holding the application  state?

It seems to me that we can just do:

```python
application_state_transition(state.application_state_root, beacon_chain_data, body.application_payload)
```
That saves 1 round trip of retrieving and passing the application state

Reply:
> Is the beacon node holding the application state?

This is implementation dependant. If there is a client that connected to both networks (consensus and application) then it has direct access to the application state. The other type of client has separated pieces of software representing consensus and application parts, these pieces are interacting via RPC protocol. This is the reason behind keeping the spec abstract in that regard.

The other reason for that is the difference between commitment schemes of beacon state and application state, that is binary tree vs MPT. If the scheme were the same we could think of _embedding_ application state into the beacon state. Then application state would be a part of the beacon state for the monolithic type of clients and could be still represented as a state root for clients of separated architecture.

Reply:
> It seems to me that we can just do:
> ```python
> application_state_transition(state.application_state_root, beacon_chain_data, body.application_payload)
> ```
> That saves 1 round trip of retrieving and passing the application state

Currently, it follows a logic similar to beacon chain `state_transition` function that accepts the `(state, block)` tuple and mutates the state by applying the state transition.